src/Workspaces/Core/Portable/Rename/Renamer.cs

```diff
@@ -25,6 +30,109 @@ public static class Renamer
+                .WithDocumentFolders(document.Id, newDocumentInfo.Folders);
+
+            if (!(shouldRenameType || shouldMoveNamespace))
+            {
+                return new RenameDocumentResult(solution, renameSymbolSucceeded: false);
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            if (shouldRenameType)
+            {
+                solution = await RenameTypeToMatchDocumentNameAsync(document.Id, solution, newName, document.Name, cancellationToken).ConfigureAwait(false);
+            }
+
+            if (shouldMoveNamespace)
+            {
+                solution = await SyncNamespaceToDocumentFoldersAsync(document.Id, solution, cancellationToken).ConfigureAwait(false);
+            }
+
+            return new RenameDocumentResult(solution, renameSymbolSucceeded: true);
+        }
+
+        private static async Task<Solution> RenameTypeToMatchDocumentNameAsync(DocumentId documentId, Solution solution, string newName, string originalName, CancellationToken cancellationToken)
+        {
+            var newNameWithoutExtension = Path.GetFileNameWithoutExtension(newName);
+            var oldNameWithoutExtension = Path.GetFileNameWithoutExtension(originalName);
+
+            var document = solution.GetDocument(documentId);
+
+            var syntaxRoot = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+
+            var typeDeclarations = syntaxRoot.DescendantNodesAndSelf().Where(syntaxFacts.IsTypeDeclaration);
+            var displayNames = typeDeclarations.Select(d => syntaxFacts.GetDisplayName(d, DisplayNameOptions.None));
+
+            var matchingTypeDeclaration = typeDeclarations.FirstOrDefault(d => syntaxFacts.GetDisplayName(d, DisplayNameOptions.None).Equals(oldNameWithoutExtension, StringComparison.OrdinalIgnoreCase));
+            if (matchingTypeDeclaration is object)
+            {
+                var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+                var symbol = semanticModel.GetDeclaredSymbol(matchingTypeDeclaration, cancellationToken);
+
+                solution = await RenameSymbolAsync(solution, symbol, newNameWithoutExtension, solution.Options, cancellationToken).ConfigureAwait(false);
+            }
+
+            return solution;
+        }
+
+        private static async Task<Solution> SyncNamespaceToDocumentFoldersAsync(DocumentId documentId, Solution solution, CancellationToken cancellationToken)
+        {
+            var document = solution.GetDocument(documentId);
+            var changeNamespaceService = document.GetRequiredLanguageService<IChangeNamespaceService>();
+
+            var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+            var targetNamespace = BuildNamespaceFromFolders(document.Folders, syntaxFacts);
```

Comment:
BuildNamespaceFromFolders can return null.  I don't see anything checking that.

Reply:
Null is valid in this case. According to `ChangeNamespaceAsync` it will move to the global namespace if possible.

Reply:
got it.  can we doc this at appropriate points in the code/xml-docs?

Reply:
Will do 👍

Reply:
🥚 > 😟 

There are code paths in sync namespace that take null as a target namespace, but this expects empty string instead. Updating